pkg/envinfo/envinfo.go

```diff
@@ -0,0 +1,281 @@
+			} else {
+				esi.componentSettings.URL = &[]ConfigURL{urlValue}
+			}
+		}
+
+		return esi.writeToFile()
+	}
+	return errors.Errorf("unknown parameter :'%s' is not a parameter in local odo config", parameter)
+
+}
+
+// DeleteEnvDirIfEmpty Deletes the env directory if its empty
+func (esi *EnvSpecificInfo) DeleteEnvDirIfEmpty() error {
+	envDir := filepath.Dir(esi.Filename)
+	_, err := esi.fs.Stat(envDir)
+	if os.IsNotExist(err) {
+		// If the Env dir doesn't exist then we dont mind
+		return nil
+	} else if err != nil {
+		// Possible to not have permission to the dir
+		return err
+	}
+	f, err := esi.fs.Open(envDir)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+	_, err = f.Readdir(1)
+
+	// If directory is empty we can remove it
+	if err == io.EOF {
+		glog.V(4).Info("Deleting the env directory as well because its empty")
+
+		return esi.fs.Remove(envDir)
+	}
+	return err
+}
+
+// DeleteEnvInfoFile deletes the envinfo.yaml file if it exists
+func (esi *EnvSpecificInfo) DeleteEnvInfoFile() error {
+	return util.DeletePath(esi.Filename)
+}
+
+// IsSet uses reflection to get the parameter from the envinfo struct, currently
+// it only searches the componentSettings
+func (esi *EnvSpecificInfo) IsSet(parameter string) bool {
+	return util.IsSet(esi.componentSettings, parameter)
+}
+
+// EnvInfoFileExists if the envinfo file exists or not
+func (esi *EnvSpecificInfo) EnvInfoFileExists() bool {
+	return esi.envinfoFileExists
+}
+
+// DeleteEnvSpecificInfo is used to delete environment specific info from local odo envinfo
+func (esi *EnvSpecificInfo) DeleteEnvSpecificInfo(parameter string) error {
+	if parameter, ok := asLocallySupportedParameter(parameter); ok {
+
+		switch parameter {
+		default:
+			if err := util.DeleteConfiguration(&esi.componentSettings, parameter); err != nil {
+				return err
+			}
+		}
+		return esi.writeToFile()
+	}
+	return errors.Errorf("unknown parameter :'%s' is not a parameter in local odo config", parameter)
+
+}
+
+// DeleteURL is used to delete environment specific info for url from local odo envinfo
+func (esi *EnvSpecificInfo) DeleteURL(parameter string) error {
+	for i, url := range *esi.componentSettings.URL {
+		if url.ClusterHost == parameter {
+			s := *esi.componentSettings.URL
+			s = append(s[:i], s[i+1:]...)
+			esi.componentSettings.URL = &s
+		}
+	}
+	return esi.writeToFile()
+}
+
+// GetComponentSettings returns the componentSettings from envinfo
+func (esi *EnvSpecificInfo) GetComponentSettings() ComponentSettings {
+	return esi.componentSettings
+}
+
+// SetComponentSettings sets the componentSettings from to the local config and writes to the file
+func (esi *EnvSpecificInfo) SetComponentSettings(cs ComponentSettings) error {
+	esi.componentSettings = cs
+	return esi.writeToFile()
+}
+
+func (esi *EnvSpecificInfo) writeToFile() error {
+	pei := newProxyEnvInfo()
+	pei.ComponentSettings = esi.componentSettings
+
+	return util.WriteToFile(&pei, esi.Filename)
```

Comment:
Please modify the util function to use a fake filesystem for faster testing

Reply:
Here we want to write to  the file on actual file system. It's not a test.

Reply:
I know but if we modify the util.WriteToFile() to take a filesystem and send the a fake one from here, the writes to the disk will be replaced with in-memory writes and thus `esi.SetConfiguration` and other similar calls will be faster in tests.